### ISSUE TITLE
[Improvement] Add params to Challenge

### ISSUE THREAD

**Issue by rfc2822** [square/okhttp#2772]
At the moment, `Challenge` only contains the auth scheme name and realm. To provide authentication other than `Basic`, this is not enough. For instance, for `Digest` auth, there are various parameters which have to be parsed and processed.

I suggest to improve the `Challenge` class so that it contains not only auth scheme name and realm, but also a list of auth parameters, so that implementors don't have to parse `WWW-Authenticate` again and can use okhttp's `Challenge` instead, which would reduce redundancy. Logically, it sounds reasonable for me that a HTTP framework prepares authentication challenges in a way that they're already usable by authenticators.

For a suggested layout of `Challenge`, please have a look at this [HttpUtils.AuthScheme](https://gitlab.com/bitfireAT/dav4android/blob/bf2004e9052241345c435c2d46090ab9baf43eb8/src/main/java/at/bitfire/dav4android/HttpUtils.java#L118). It contains the auth scheme name, parsed unnamed parameters and parsed named parameters.

For a suggested parsing algorithm of `WWW-Authenticate` that generates such parameterized challenge instances, please see [HttpUtils.parseWwwAuthenticate](https://gitlab.com/bitfireAT/dav4android/blob/bf2004e9052241345c435c2d46090ab9baf43eb8/src/main/java/at/bitfire/dav4android/HttpUtils.java#L32) and its [tests](https://gitlab.com/bitfireAT/dav4android/blob/bf2004e9052241345c435c2d46090ab9baf43eb8/src/androidTest/java/at/bitfire/dav4android/HttpUtilsTest.java#L17). It handles challenges without parameters, with RFC-compliant parameters (`param=value`), and with unnamed parameters (like used in `Negotiate`).

**Comment by swankjesse** [comment]
Is this for digest authentication? Consider using this project:
https://github.com/rburgst/okhttp-digest

At the moment we have no specific plans to offer digest authentication in OkHttp itself.

**Comment by rfc2822** [comment]
It's not only for Digest authentication, but for *any* authentication that requires parameters (i.e. any except `Basic` auth). I have already done a [working Basic/Digest implementation](https://gitlab.com/bitfireAT/dav4android/blob/master/src/main/java/at/bitfire/dav4android/BasicDigestAuthHandler.java) (without strange imported code from Apache HttpClient). I just think

* it's unnecessary overhead to parse `WWW-Authenticate` headers multiple times (one time in `Challenge`, and then the authenticator/authentication interceptor has to parse everything again),
* and that it be in the scope of an HTTP framework to parse and prepare the authentication headers for its authenticators (especially because every authenticator would have to implement the same redundant parsing code itself, as it's a HTTP thing).

However, it seems that you have another opinion, which is absolutely OK. Thanks for your great (and free) work.

**Comment by swankjesse** [comment]
Cool! Maybe the thing I’m missing is that OkHttp doesn’t attempt to implement authentication for arbitrary servers. We assume that callers know if they need basic vs. digest, whereas dav4android doesn’t make that assumption.

Wanna submit the enhancements?

**Comment by rfc2822** [comment]
Yes, the dav4android methods are designed to be independent of the actual authentication method. 

To implement my suggestion, I'd
* enhance `Challenge` to also store unnamed and named parameters so that authenticators can use `response.challanges()` to get all required information without parsing headers itself;
* change the `List<Challenge> parseChallenges(Headers responseHeaders, String challengeHeader)` method so that it extracts name *and parameters* from `challengeHeader`.

**Comment by alexeyvasilyev** [comment]
I also agree that `Challenge` class should have a set of parameters (not only `scheme` and `realm`).

**Comment by murki** [comment]
Seems that as of v3.8 this hasn't been implemented. Should the milestone be updated or am I missing something?
